### DISCUSSION TITLE
How to round up the light year

### DISCUSSION THREAD

**Discussion by flawrist** [PaddiM8/kalk#131] [Q&A] [answered]
Dear fellows,
can you tell me how to express the light year shorter?

Right now there is

``` console
$ kalker 9460730472581
9 460 730 472 581 ≈ 9.4607304726×10^12
```

What I'm looking for is `9.46×10^12` or `9.46e12`.

**Comment by PaddiM8** [comment]
`9.46×10^12` should work, no?

**Reply by flawrist** [reply]
I believe the question consists of two parts:
* how to round/truncate/whatever 9460730472581 to 9460000000000
* how to change the notation of 9460000000000 to 9.46×10^12 or 9.46e12

**Reply by PaddiM8** [reply]
It can be rounded by doing `round(9460730472581/(10^10))*10^10` where ^10 is the amount of digits that should be rounded. I guess you could automatically figure out the digit count with the help of the `log` function as well. Not sure what you mean with the second one since you get `9.46×10^12` when entering `9460000000000`

**Reply by flawrist** [reply]
I enter not 9460000000000, but 9460730472581, and get 9.4607304726×10^12 instead of 9.46e12. If I use [Qalc](https://github.com/Qalculate/libqalculate), which is a bit laggy on my end (that's why I use Kalker), there is an easy way to do that:
``` console
$ qalc.exe -s "exp 1; maxdeci 2; lowe" 1 ly to km
1 lightyear = approx. 9,46e12 km
```
where `exp 1` switches on E notation, `maxdeci 2` truncates given number, `lowe` changes E to e. 

So I wonder if there is an easy way to do the same with Kalker.
You mentioned `round(9460730472581/(10^10))*10^10`
which looks too long, counter-intuitive and scary, and said nothing
about any options to activate E notation (perhaps it is not supported).

**Reply by PaddiM8** [reply]
Hmm I see. E notation is unfortunately not supported, since it would be parsed as `9.46*e*12`, where e is the variable e. A function to truncate decimals doesn't exist, but it would be a useful function to have, so I might add it in the future! For now, you could create your own, by putting it in the `~/.config/kalker/` file (or `%appdata%/kalker/` on Windows).

truncate(12.3456, 2) => 12.34 would be:
```
truncate(x, n) = floor(x * 10^n) / 10^n
```

After doing this, the function should always be available.

**Reply by flawrist** [reply]
Sorry, what file should I create in this path `%appdata%/kalker/` you mentioned?
Now I have nothing there, or rather, there is `history.txt` and that’s it.

Perhaps you mean there should be some configuration file created after installation, but you see, I'm not installing Kalker via MSI, After downloading MSI, I extract Kalker using [Lessmsi](https://github.com/activescott/lessmsi) by @activescott and put it in a folder with other portable apps (let's say C:\APPS\PORTABLE) added to PATH variable.

**Reply by PaddiM8** [reply]
Oops, sorry, forgot to mention the file name. `%appdata%/kalker/default.kalker` is the file. `history.txt` being there is a good sign.

**Reply by flawrist** [reply]
``` console
$ type default.kalker
truncate(x, n) = floor(x * 10^n) / 10^n

$ kalker truncate(9460730472581, 2)
9 460 730 472 581 ≈ 9.4607304726×10^12
```
whereas `9.46×10^12` is expected.

**Reply by PaddiM8** [reply, marked as answer]
Ah well for integers it would be `truncate(x, n) = floor(x / 10^(floor(log(x)) - n)) * 10^floor(log(x) - n)` I think. A bit more complicated, but `10^(floor(log(x)) - n)` could be its own function to make it more readable.

**Reply by flawrist** [reply]
Congratulations, @PaddiM8, we did it.
By the way, Qalc comes from Sweden as well, @hanna-kn is [especially](https://github.com/Qalculate/libqalculate/issues/5#issuecomment-825571565) [helpful](https://github.com/Qalculate/libqalculate/issues/149#issuecomment-1898667340).

**Reply by PaddiM8** [reply]
Ohh nice, that was fun. Didn't know that about Qalc, neat.
